sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PartCreationPolicyShouldBeUsedWithExportAttribute.cs

```diff
@@ -0,0 +1,78 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2018 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class PartCreationPolicyShouldBeUsedWithExportAttribute : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4428";
+        private const string MessageFormat = "Add the 'ExportAttribute' or remove 'PartCreationPolicyAttribute' to/from this class definition.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<KnownType> ExportAttributes = new HashSet<KnownType>
+        {
+            KnownType.System_ComponentModel_Composition_ExportAttribute,
+            KnownType.System_ComponentModel_Composition_InheritedExportAttribute,
+        };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var attribute = (AttributeSyntax)c.Node;
+
+                    if (IsPartCreationPolicyAttribute(attribute) &&
+                        attribute.FirstAncestorOrSelf<MemberDeclarationSyntax>() is ClassDeclarationSyntax classDeclaration)
+                    {
+                        var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+                        if (classSymbol != null &&
+                            !classSymbol.GetAttributes().Any(IsExportAttribute))
```

Comment:
I don't think this is the right check. AFAIK the goal of `InheritedExport` is to say you don't have to specify the `Export` on sub-classes so your rule would raise an issue where it shouldn't. You will have to walk up the tree looking for `InheritedExport`.